### PR TITLE
Add retry_policy for graphs

### PR DESCRIPTION
**Pull request by nfcampos**
- default is to not retry, fully configurable
- configuration options follow temporal https://docs.temporal.io/retry-policies#properties

### CODE REVIEW

File: langgraph/pregel/__init__.py [hunk 1]
```diff
@@ -487,7 +492,9 @@ def update_state(
         saved = self.checkpointer.get_tuple(config)
         checkpoint = copy_checkpoint(saved.checkpoint) if saved else empty_checkpoint()
         # find last node that updated the state, if not provided
-        if as_node is None and not saved:
+        if as_node is None and not any(
+            v for vv in checkpoint["versions_seen"].values() for v in vv.values()
```
Comment (hinthornw) [inline langgraph/pregel/__init__.py:496]: is this related?

File: langgraph/pregel/retry.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    ):
+        return False
+    if isinstance(exc, httpx.HTTPStatusError):
+        return 500 <= exc.response.status_code < 600
+    if isinstance(exc, requests.HTTPError):
+        return 500 <= exc.response.status_code < 600 if exc.response else True
+    return True
+
+
+class RetryPolicy(NamedTuple):
```
Comment (hinthornw) [inline langgraph/pregel/retry.py:37]: ooc: why a named tuple vs a typeddict?

File: langgraph/pregel/retry.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    while True:
+        try:
+            # clear any writes from previous attempts
+            task.writes.clear()
+            # run the task
+            if stream:
+                async for _ in task.proc.astream(task.input, task.config):
+                    pass
+            else:
+                await task.proc.ainvoke(task.input, task.config)
```
Comment (hinthornw) [inline langgraph/pregel/retry.py:114]: Could be nice to log the attempt number as metadata so you could track e.g., frequency of retries in ls
Would also presumably make it easier in something like stream events to check if you need to overwrite something? not sure

File: langgraph/pregel/retry.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            # if successful, end
+            break
+        except Exception as exc:
```
Comment (hinthornw) [inline langgraph/pregel/retry.py:117]: why not catch teh broader `BaseException` here?

**Review by hinthornw — APPROVED** [review]

File: langgraph/pregel/__init__.py (hunk 1 above)
Comment (nfcampos) [reply langgraph/pregel/__init__.py:496]: No, just an edge case I found while migrating js

File: langgraph/pregel/retry.py (hunk 2 above)
Comment (nfcampos) [reply langgraph/pregel/retry.py:37]: because it allows defaults

File: langgraph/pregel/retry.py (hunk 4 above)
Comment (nfcampos) [reply langgraph/pregel/retry.py:117]: Isn't the only difference that it includes KeyboardInterrupt and friends? I don't think we'd want to ever retry on those

File: langgraph/pregel/retry.py (hunk 3 above)
Comment (nfcampos) [reply langgraph/pregel/retry.py:114]: yea agree, but not sure today where to log it, do you have an idea?
